File: backend/routers/recovery.py
from fastapi import APIRouter, Query
from backend.database import get_db, fetch_all, fetch_one
from backend.services import dashboard as dash_svc
from backend.services.project_filter import parse_project_ids, sql_in


def _due_soon(conn, project_ids: list[int] | None) -> list[dict]:
    clause, params = sql_in("u.project_id", project_ids)
    rows = fetch_all(
        conn,
        f"""SELECT i.id, i.booking_id, i.customer_id, i.unit_id, i.amount, i.remaining_amount,
                   i.paid_amount, i.due_date, i.type, i.notes, i.status,
                   i.trigger_kind, i.trigger_label, i.trigger_progress,
                   bk.booking_no,
                   c.name AS customer_name, c.contact_number AS phone, c.cnic,
                   u.unit_no, p.name AS project_name, u.project_id,
                   CAST(julianday(i.due_date) - julianday('now') AS INT) AS days_until,
                   (SELECT MAX(py.payment_date) FROM payments py WHERE py.booking_id=i.booking_id) AS last_payment
            FROM installments i
            JOIN bookings bk ON bk.id=i.booking_id AND bk.status='active'
            JOIN customers c ON c.id=i.customer_id
            JOIN units u ON u.id=i.unit_id
            JOIN projects p ON p.id=u.project_id
            WHERE i.status IN ('pending','partial') AND i.remaining_amount > 0
              AND i.due_date >= date('now')
              AND i.due_date < date('now','start of month','+1 month'){clause}
            ORDER BY i.due_date, c.name""",
        params,
    )
    for r in rows:
        r["original_amount"] = r.get("amount")
        r["amount"] = r.get("remaining_amount") or r["amount"]
        dash_svc.explain_installment(r, overdue=False)
    return rows

router = APIRouter(prefix="/api", tags=["recovery"])


def _ageing(overdue: list[dict]) -> dict:
    buckets = {
        "d30": {"count": 0, "amount": 0},
        "d60": {"count": 0, "amount": 0},
        "d90": {"count": 0, "amount": 0},
    }
    for row in overdue:
        days = row.get("days_overdue") or 0
        amt = row.get("amount") or 0
        if days <= 30:
            key = "d30"
        elif days <= 60:
            key = "d60"
        else:
            key = "d90"
        buckets[key]["count"] += 1
        buckets[key]["amount"] += amt
    return buckets


def _recovery(conn, project_ids: list[int] | None):
    overdue = dash_svc.get_overdue_list(conn, project_ids)
    if project_ids:
        ph = ",".join("?" * len(project_ids))
        totals = fetch_one(
            conn,
            f"""SELECT COALESCE(SUM(i.remaining_amount),0) AS receivable,
                      COALESCE(SUM(CASE WHEN i.status IN ('overdue','partial') AND i.due_date < date('now') THEN i.remaining_amount ELSE 0 END),0) AS overdue_amt
               FROM installments i
               JOIN units u ON u.id=i.unit_id
               JOIN bookings bk ON bk.id=i.booking_id AND bk.status='active'
               WHERE i.status IN ('pending','partial','overdue')
                 AND u.project_id IN ({ph})""",
            tuple(project_ids),
        )
        collected = fetch_one(
            conn,
            f"""SELECT COALESCE(SUM(p.amount),0) AS v FROM payments p
               JOIN bookings b ON b.id=p.booking_id
               WHERE p.payment_date >= date('now','-30 days')
                 AND b.project_id IN ({ph})""",
            tuple(project_ids),
        )
    else:
        totals = fetch_one(
            conn,
            """SELECT COALESCE(SUM(i.remaining_amount),0) AS receivable,
                      COALESCE(SUM(CASE WHEN i.status IN ('overdue','partial') AND i.due_date < date('now') THEN i.remaining_amount ELSE 0 END),0) AS overdue_amt
               FROM installments i JOIN bookings bk ON bk.id=i.booking_id AND bk.status='active'
               WHERE i.status IN ('pending','partial','overdue')""",
        )
        collected = fetch_one(
            conn,
            "SELECT COALESCE(SUM(amount),0) AS v FROM payments WHERE payment_date >= date('now','-30 days')",
        )
    return {
        "overdue": overdue,
        "due_soon": _due_soon(conn, project_ids),
        "receivable": totals["receivable"] if totals else 0,
        "overdue_amt": totals["overdue_amt"] if totals else 0,
        "collected_month": collected["v"] if collected else 0,
        "ageing": _ageing(overdue),
    }


@router.get("/recovery")
def recovery(
    project_id: int | None = Query(None),
    project_ids: str | None = Query(None),
):
    ids = parse_project_ids(project_ids, project_id)
    with get_db() as conn:
        return _recovery(conn, ids)


@router.get("/recovery/calendar")
def recovery_calendar(
    year: int = Query(...),
    month: int = Query(...),
    project_id: int | None = Query(None),
    project_ids: str | None = Query(None),
):
    ids = parse_project_ids(project_ids, project_id)
    start = f"{year:04d}-{month:02d}-01"
    if month == 12:
        end = f"{year + 1:04d}-01-01"
    else:
        end = f"{year:04d}-{month + 1:02d}-01"
    clause, params = sql_in("u.project_id", ids)
    q = f"""SELECT i.id, i.booking_id, i.customer_id, i.due_date, i.amount,
                   i.remaining_amount AS amount_due, i.status, i.type,
                   c.name AS customer_name, u.unit_no, p.name AS project_name,
                   c.contact_number AS phone
            FROM installments i
            JOIN bookings bk ON bk.id=i.booking_id AND bk.status='active'
            JOIN customers c ON c.id=i.customer_id
            JOIN units u ON u.id=i.unit_id
            JOIN projects p ON p.id=u.project_id
            WHERE i.status IN ('pending','partial','overdue')
              AND i.due_date>=? AND i.due_date<?{clause}
            ORDER BY i.due_date, c.name"""
    with get_db() as conn:
        rows = fetch_all(conn, q, (start, end) + params)
    days: dict = {}
    for r in rows:
        day = (r.get("due_date") or "")[:10]
        bucket = days.setdefault(day, {"date": day, "count": 0, "amount": 0, "items": []})
        due = r.get("amount_due") or r.get("amount") or 0
        bucket["count"] += 1
        bucket["amount"] += due
        r["amount"] = due
        bucket["items"].append(r)
    return {"year": year, "month": month, "days": list(days.values())}


@router.get("/demand-notices")
def demand_notices(
    project_id: int | None = Query(None),
    project_ids: str | None = Query(None),
):
    ids = parse_project_ids(project_ids, project_id)
    with get_db() as conn:
        return dash_svc.get_overdue_list(conn, ids)

File: static/js/screens/recovery.js
import { $, esc } from '../dom.js';
import { api, toast } from '../api.js';
import { fmt, fmtShort, overdueBadge } from '../format.js';
import { projectFilterQuery } from '../project-filter.js';
import { openModal, closeModal } from '../modal.js';
import { getCurrentScreen, reloadCurrentScreen } from '../nav.js';
import { openCustomerDetail } from './customers.js';
import { loadDashboard } from './dashboard.js';
import { askConfirm } from '../dialog.js';

let recoveryOverdue = [];
let recoverySoon = [];
let recTab = 'overdue';
let payChoices = [];
let ageFilter = null;
let payContext = null;
let calYear = null;
let calMonth = null;
let calDays = [];
let calSelected = null;

function todayISO() {
  const d = new Date();
  const p = (n) => String(n).padStart(2, '0');
  return `${d.getFullYear()}-${p(d.getMonth() + 1)}-${p(d.getDate())}`;
}

function ageBucket(days) {
  if (days <= 30) return 'd30';
  if (days <= 60) return 'd60';
  return 'd90';
}

function ageingFromRows() {
  const out = {
    d30: { count: 0, amount: 0 },
    d60: { count: 0, amount: 0 },
    d90: { count: 0, amount: 0 },
  };
  recoveryOverdue.forEach((o) => {
    const key = ageBucket(o.days_overdue || 0);
    out[key].count += 1;
    out[key].amount += o.amount || 0;
  });
  return out;
}

function caseBlob(o) {
  return [o.customer_name, o.unit_no, o.project_name, o.phone, o.cnic, o.type, o.trigger_label, o.reason, o.why]
    .filter(Boolean).join(' ').toLowerCase();
}

function filteredCases() {
  const q = ($('rec-search')?.value || '').trim().toLowerCase();
  const source = recTab === 'soon' ? recoverySoon : recoveryOverdue;
  return source.filter((o) => {
    if (recTab === 'overdue' && ageFilter && ageBucket(o.days_overdue || 0) !== ageFilter) return false;
    if (!q) return true;
    return caseBlob(o).includes(q);
  });
}

export async function loadRecovery() {
  const d = await api(`/api/recovery${projectFilterQuery()}`);
  recoveryOverdue = Array.isArray(d.overdue) ? d.overdue : [];
  recoverySoon = Array.isArray(d.due_soon) ? d.due_soon : [];
  if ($('r-recv')) $('r-recv').textContent = fmtShort(d.receivable);
  if ($('r-over')) $('r-over').textContent = fmtShort(d.overdue_amt);
  if ($('r-coll')) $('r-coll').textContent = fmtShort(d.collected_month);
  if ($('r-cases')) $('r-cases').textContent = recoveryOverdue.length;
  paintAgeing(d.ageing && d.ageing.d30 ? d.ageing : ageingFromRows());
  syncRecTabs();
  renderRecoveryTable();
  await loadCalendar();
}

function ensureCalMonth() {
  if (calYear && calMonth) return;
  const n = new Date();
  calYear = n.getFullYear();
  calMonth = n.getMonth() + 1;
}

async function loadCalendar() {
  if (!$('rec-calendar')) return;
  ensureCalMonth();
  const extra = projectFilterQuery();
  const url = `/api/recovery/calendar?year=${calYear}&month=${calMonth}${extra ? `&${extra.slice(1)}` : ''}`;
  try {
    const d = await api(url);
    calDays = d.days || [];
  } catch {
    calDays = [];
  }
  renderCalendar();
}

function renderCalendar() {
  const grid = $('rec-calendar');
  if (!grid) return;
  const label = $('cal-label');
  if (label) {
    label.textContent = new Date(calYear, calMonth - 1, 1).toLocaleString('en-GB', {
      month: 'long', year: 'numeric',
    });
  }
  const byDate = {};
  calDays.forEach((d) => { byDate[d.date] = d; });
  const first = new Date(calYear, calMonth - 1, 1);
  const startDow = first.getDay();
  const daysInMonth = new Date(calYear, calMonth, 0).getDate();
  const today = todayISO();
  const dow = ['Sun', 'Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat'];
  let html = dow.map((n) => `<div class="cal-dow">${n}</div>`).join('');
  for (let i = 0; i < startDow; i += 1) html += '<div class="cal-cell empty"></div>';
  for (let day = 1; day <= daysInMonth; day += 1) {
    const iso = `${calYear}-${String(calMonth).padStart(2, '0')}-${String(day).padStart(2, '0')}`;
    const bucket = byDate[iso];
    const cls = [
      'cal-cell',
      bucket ? 'has-due' : '',
      iso === today ? 'today' : '',
      calSelected === iso ? 'today' : '',
    ].filter(Boolean).join(' ');
    html += `<button type="button" class="${cls}" data-cal-day="${iso}">
      <div class="cal-day">${day}</div>
      ${bucket ? `<div class="cal-meta">${bucket.count} · ${fmtShort(bucket.amount)}</div>` : ''}
    </button>`;
  }
  grid.innerHTML = html;
  grid.querySelectorAll('[data-cal-day]').forEach((btn) => {
    btn.addEventListener('click', () => showCalDay(btn.dataset.calDay));
  });
  if (calSelected) showCalDay(calSelected);
  else if ($('cal-day-list')) $('cal-day-list').hidden = true;
}

function showCalDay(iso) {
  calSelected = iso;
  const box = $('cal-day-list');
  if (!box) return;
  const bucket = calDays.find((d) => d.date === iso);
  if (!bucket || !bucket.items?.length) {
    box.hidden = true;
    box.innerHTML = '';
    return;
  }
  box.hidden = false;
  box.innerHTML = `
    <div style="font-weight:800;font-size:13px;margin-bottom:8px">Due ${esc(iso)}</div>
    <div class="tbl-wrap"><table>
      <thead><tr><th>Customer</th><th>Unit</th><th>Type</th><th>Due</th><th></th></tr></thead>
      <tbody>${bucket.items.map((i) => `
        <tr>
          <td class="td-b">${esc(i.customer_name)}</td>
          <td>${esc(i.unit_no)}</td>
          <td>${esc(i.type || '—')}</td>
          <td class="td-red">${fmt(i.amount)}</td>
          <td><button type="button" class="btn sm primary" data-cal-pay="${i.id}">Pay</button></td>
        </tr>`).join('')}
      </tbody>
    </table></div>`;
  box.querySelectorAll('[data-cal-pay]').forEach((btn) => {
    btn.addEventListener('click', () => {
      const row = bucket.items.find((x) => x.id === parseInt(btn.dataset.calPay, 10));
      if (row) openPayForInstallment(row);
    });
  });
}

function paintAgeing(ageing) {
  ['d30', 'd60', 'd90'].forEach((key) => {
    const b = ageing[key] || { count: 0, amount: 0 };
    if ($(`age-${key}-amt`)) $(`age-${key}-amt`).textContent = fmtShort(b.amount);
    if ($(`age-${key}-cnt`)) $(`age-${key}-cnt`).textContent = `${b.count} case${b.count === 1 ? '' : 's'}`;
  });
  document.querySelectorAll('.age-card').forEach((el) => {
    el.classList.toggle('active', el.dataset.age === ageFilter);
  });
}

function findCase(id) {
  return recoveryOverdue.find((x) => x.id === id)
    || recoverySoon.find((x) => x.id === id)
    || payChoices.find((x) => x.id === id)
    || null;
}

function syncRecTabs() {
  document.querySelectorAll('[data-rec-tab]').forEach((btn) => {
    btn.classList.toggle('active', btn.dataset.recTab === recTab);
  });
  const hint = $('rec-tab-hint');
  if (hint) {
    hint.textContent = recTab === 'soon'
      ? `${recoverySoon.length} installment${recoverySoon.length === 1 ? '' : 's'} still upcoming this month`
      : `${recoveryOverdue.length} overdue installment${recoveryOverdue.length === 1 ? '' : 's'}`;
  }
  document.querySelectorAll('.age-card').forEach((el) => {
    el.disabled = recTab === 'soon';
    el.classList.toggle('is-dim', recTab === 'soon');
  });
}

function renderRecoveryTable() {
  const rows = filteredCases();
  const tbody = $('recovery-tbody');
  if (!tbody) return;
  const empty = recTab === 'soon'
    ? (recoverySoon.length ? 'No matching upcoming installments' : 'Nothing else due this month')
    : (recoveryOverdue.length ? 'No matching installments' : 'No overdue installments');
  tbody.innerHTML = rows.length
    ? rows.map((o) => {
        const days = o.days_overdue || 0;
        const bucket = recTab === 'overdue' ? ageBucket(days) : '';
        const when = o.when || (recTab === 'overdue'
          ? `${days}d overdue · ${o.due_date || ''}`
          : o.due_date || '—');
        return `
      <tr class="${bucket ? `rec-${bucket}` : ''}">
        <td>
          <div class="td-b">${esc(o.customer_name)}</div>
          <div class="td-sm">${esc(o.phone || '—')}</div>
        </td>
        <td>
          <div>${esc(o.unit_no)}</div>
          <div class="td-sm">${esc(o.project_name || '')}</div>
        </td>
        <td class="rec-why">
          <div>${esc(o.why || o.trigger_label || o.type || 'Installment')}</div>
          <div class="td-sm">${esc(o.what || '')}</div>
        </td>
        <td class="${recTab === 'overdue' ? 'td-red' : ''}">${fmt(o.amount)}</td>
        <td>${recTab === 'overdue'
          ? `<span class="badge ${overdueBadge(days)}">${esc(when)}</span>`
          : esc(when)}</td>
        <td style="white-space:nowrap">
          <button type="button" class="btn sm primary" data-rec-pay="${o.id}">Pay</button>
          ${recTab === 'overdue' ? `<button type="button" class="btn sm" data-rec-notice="${o.id}">Notice</button>` : ''}
          <button type="button" class="btn sm" data-rec-view="${o.customer_id}">View</button>
        </td>
      </tr>`;
      }).join('')
    : `<tr><td colspan="6" style="text-align:center;color:var(--g400);padding:20px">${empty}</td></tr>`;

  tbody.querySelectorAll('[data-rec-view]').forEach((btn) => {
    btn.addEventListener('click', () => openCustomerDetail(parseInt(btn.dataset.recView, 10)));
  });
  tbody.querySelectorAll('[data-rec-pay]').forEach((btn) => {
    btn.addEventListener('click', () => {
      const row = findCase(parseInt(btn.dataset.recPay, 10));
      if (row) openPayModal(row);
    });
  });
  tbody.querySelectorAll('[data-rec-notice]').forEach((btn) => {
    btn.addEventListener('click', async () => {
      const row = findCase(parseInt(btn.dataset.recNotice, 10));
      if (!row) return;
      const { openGenerate } = await import('./documents.js');
      await openGenerate({ customerId: row.customer_id, bookingId: row.booking_id, kind: 'notice' });
    });
  });
}

function resetPayForm() {
  payContext = null;
  ['pay-inst-id', 'pay-bk-id', 'pay-cust-id', 'pay-amount', 'pay-ref', 'pay-notes'].forEach((id) => {
    if ($(id)) $(id).value = '';
  });
  if ($('pay-method')) $('pay-method').value = 'Cash';
  if ($('pay-date')) $('pay-date').value = todayISO();
  if ($('pay-summary')) {
    $('pay-summary').hidden = true;
    $('pay-summary').innerHTML = '';
  }
  if ($('pay-pick-wrap')) $('pay-pick-wrap').hidden = true;
  if ($('pay-pick')) $('pay-pick').innerHTML = '<option value="">Select…</option>';
}

function fillPayFromRow(row) {
  $('pay-inst-id').value = String(row.id);
  $('pay-bk-id').value = String(row.booking_id);
  $('pay-cust-id').value = String(row.customer_id);
  $('pay-amount').value = String(row.amount || row.remaining_amount || '');
  $('pay-summary').hidden = false;
  $('pay-summary').innerHTML = `
    <div class="bk-dname">${esc(row.customer_name)}</div>
    <div class="sum-row"><span class="sum-lbl">Unit</span><span class="sum-val">${esc(row.unit_no)} · ${esc(row.project_name)}</span></div>
    <div class="sum-row"><span class="sum-lbl">Why</span><span class="sum-val">${esc(row.why || row.trigger_label || row.type || 'Installment')}</span></div>
    <div class="sum-row"><span class="sum-lbl">When</span><span class="sum-val">${esc(row.when || row.due_date || '—')}</span></div>
    <div class="sum-row"><span class="sum-lbl">Remaining</span><span class="sum-val td-red">${fmt(row.amount)}</span></div>`;
}

export function openPayForInstallment(row) {
  resetPayForm();
  payContext = row;
  fillPayFromRow(row);
  openModal('pay-modal');
}

async function loadPayChoices() {
  const seen = new Map();
  recoveryOverdue.forEach((o) => seen.set(o.id, o));
  const now = new Date();
  for (let i = 0; i < 3; i++) {
    let y = now.getFullYear();
    let m = now.getMonth() + 1 + i;
    if (m > 12) { m -= 12; y += 1; }
    try {
      const cal = await api(`/api/recovery/calendar?year=${y}&month=${m}${projectFilterQuery()}`);
      for (const day of cal.days || []) {
        for (const item of day.items || []) {
          if (!seen.has(item.id)) seen.set(item.id, item);
        }
      }
    } catch { /* ignore */ }
  }
  return [...seen.values()];
}

async function openPayModal(row = null) {
  resetPayForm();
  openModal('pay-modal');
  if (row) {
    payContext = row;
    fillPayFromRow(row);
    return;
  }
  payChoices = await loadPayChoices();
  if (!payChoices.length) {
    closeModal('pay-modal');
    toast('No upcoming or overdue installments.', 'error');
    return;
  }
  $('pay-pick-wrap').hidden = false;
  $('pay-pick').innerHTML = '<option value="">Select…</option>' + payChoices.map((o) =>
    `<option value="${o.id}">${esc(o.customer_name)} · ${esc(o.unit_no)} · ${fmt(o.amount || o.amount_due || 0)}</option>`).join('');
}

function onPayPickChange() {
  const id = parseInt($('pay-pick')?.value, 10);
  const row = payChoices.find((x) => x.id === id) || recoveryOverdue.find((x) => x.id === id);
  if (row) {
    payContext = row;
    fillPayFromRow(row);
  }
  else {
    $('pay-summary').hidden = true;
    ['pay-inst-id', 'pay-bk-id', 'pay-cust-id', 'pay-amount'].forEach((i) => { if ($(i)) $(i).value = ''; });
  }
}

async function submitPayment() {
  const instId = parseInt($('pay-inst-id').value, 10);
  const bkId = parseInt($('pay-bk-id').value, 10);
  const custId = parseInt($('pay-cust-id').value, 10);
  const amount = parseInt($('pay-amount').value, 10);
  const row = findCase(instId) || (payContext && payContext.id === instId ? payContext : null);
  const due = row?.amount || row?.remaining_amount || 0;
  if (!instId || !bkId || !custId) {
    toast('Select an installment.', 'error');
    return;
  }
  if (!amount || amount < 1) {
    toast('Enter a payment amount.', 'error');
    return;
  }
  if (due && amount > due) {
    toast('Amount cannot exceed remaining due.', 'error');
    return;
  }
  const who = row ? `${row.customer_name} / ${row.unit_no}` : 'this installment';
  if (!await askConfirm(`Record payment of ${fmt(amount)} for ${who}?`, {
    title: 'Record payment',
    confirmLabel: 'Record payment',
  })) return;
  try {
    const r = await api('/api/payments', {
      method: 'POST',
      body: JSON.stringify({
        installment_id: instId,
        booking_id: bkId,
        customer_id: custId,
        amount,
        payment_date: $('pay-date').value || todayISO(),
        method: $('pay-method').value || 'Cash',
        reference_number: $('pay-ref').value.trim() || null,
        notes: $('pay-notes').value.trim() || null,
      }),
    });
    closeModal('pay-modal');
    toast(`Payment recorded · ${r.receipt}`);
    await loadRecovery();
    try { await loadDashboard(); } catch { /* badge refresh is best-effort */ }
    if (getCurrentScreen() === 'portal') reloadCurrentScreen();
  } catch {
    /* api() already toasts */
  }
}

export function initRecoveryEvents() {
  $('rec-search')?.addEventListener('input', renderRecoveryTable);
  document.querySelectorAll('[data-rec-tab]').forEach((btn) => {
    btn.addEventListener('click', () => {
      recTab = btn.dataset.recTab || 'overdue';
      if (recTab === 'soon') ageFilter = null;
      syncRecTabs();
      paintAgeing(ageingFromRows());
      renderRecoveryTable();
    });
  });
  $('btn-rec-pay')?.addEventListener('click', () => { openPayModal(null); });
  $('pay-pick')?.addEventListener('change', onPayPickChange);
  $('btn-save-payment')?.addEventListener('click', submitPayment);
  $('cal-prev')?.addEventListener('click', () => {
    ensureCalMonth();
    calMonth -= 1;
    if (calMonth < 1) { calMonth = 12; calYear -= 1; }
    calSelected = null;
    loadCalendar();
  });
  $('cal-next')?.addEventListener('click', () => {
    ensureCalMonth();
    calMonth += 1;
    if (calMonth > 12) { calMonth = 1; calYear += 1; }
    calSelected = null;
    loadCalendar();
  });
  document.querySelectorAll('.age-card').forEach((el) => {
    el.addEventListener('click', () => {
      const key = el.dataset.age;
      ageFilter = ageFilter === key ? null : key;
      paintAgeing(ageingFromRows());
      renderRecoveryTable();
    });
  });
}
